# Fase 1 · Análisis Exploratorio de Datos (EDA)
 
## Objetivo
 
El objetivo de esta fase es realizar una exploración inicial del dataset **DataCo Smart Supply Chain**, el cual simula las operaciones comerciales, logísticas y de cadena de suministro de una organización global. Esta revisión permite comprender la estructura multidimensional de los datos, evaluar su calidad y detectar posibles inconsistencias o sesgos que puedan afectar a las fases posteriores de modelado predictivo y analítica avanzada.
 
En esta etapa nos centramos en:
 
- **Identificar las dimensiones operativas:** Clasificar y entender las variables comerciales (ventas, clientes), logísticas (modos de envío, tiempos de entrega) y de seguridad (estados de órdenes y alertas de fraude).
- **Analizar la consistencia estructural:** Evaluar las dimensiones del dataset, los tipos de datos (fechas, variables categóricas, numéricas) y la coherencia interna entre registros relacionados.
- **Evaluar la calidad de los datos:** Detectar la presencia de valores nulos, registros duplicados, inconsistencias en las coordenadas geográficas y datos potencialmente sensibles o expuestos.
- **Explorar variables críticas del negocio:** Analizar el comportamiento de métricas clave como los días de envío reales frente a los estimados (`SLA`), los patrones de fraude en los métodos de pago y la distribución de ventas por regiones.
- **Anticipar necesidades de ingeniería de características:** Identificar requerimientos específicos de limpieza, normalización de columnas, transformación de variables temporales y preparación de datos para la fase ETL.
 
Este análisis proporciona una base sólida para definir la estrategia de preparación de datos, asegurando un pipeline robusto para la toma de decisiones estratégicas en la cadena de suministro.

In [2]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import warnings
import sys
import os # Esto sube un nivel en las carpetas para encontrar la raíz del proyecto
sys.path.append(os.path.abspath(os.path.join('..')))
import src.exploration as e

In [3]:
warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

In [4]:
df_data = pd.read_csv("../files/raw/dataco_supply_chain.csv", sep=None, engine="python", encoding="latin-1")

In [5]:
e.info_df(df_data)


INFORME TÉCNICO DEL DATAFRAME

🔹 Dimensiones:
- Filas: 180519
- Columnas: 53

🔹 Info general:
<class 'pandas.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  str    
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  str    
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  str    
 9   Customer City                  180519 non-null  str    
 10  Customer Country               180519 non-null  str    
 11  Custome

,count,unique,top,freq
Type,180519,4,DEBIT,69295
Delivery Status,180519,4,Late delivery,98977
Category Name,180519,50,Cleats,24551
Customer City,180519,563,Caguas,66770
Customer Country,180519,2,EE. UU.,111146
Customer Email,180519,1,XXXXXXXXX,180519
Customer Fname,180519,782,Mary,65150
Customer Lname,180511,1109,Smith,64104
Customer Password,180519,1,XXXXXXXXX,180519
Customer Segment,180519,3,Consumer,93504



🔹 Descriptivo variables cuantitativas:


,count,mean,std,min,25%,50%,75%,max
Days for shipping (real),180519.0,3.50,1.62,0.00,2.00,3.00,5.00,6.00
Days for shipment (scheduled),180519.0,2.93,1.37,0.00,2.00,4.00,4.00,4.00
Benefit per order,180519.0,21.97,104.43,-4274.98,7.00,31.52,64.80,911.80
Sales per customer,180519.0,183.11,120.04,7.49,104.38,163.99,247.40,1939.99
Late_delivery_risk,180519.0,0.55,0.50,0.00,0.00,1.00,1.00,1.00
Category Id,180519.0,31.85,15.64,2.00,18.00,29.00,45.00,76.00
Customer Id,180519.0,6691.38,4162.92,1.00,3258.50,6457.00,9779.00,20757.00
Customer Zipcode,180516.0,35921.13,37542.46,603.00,725.00,19380.00,78207.00,99205.00
Department Id,180519.0,5.44,1.63,2.00,4.00,5.00,7.00,12.00
Latitude,180519.0,29.72,9.81,-33.94,18.27,33.14,39.28,48.78



🔹 Muestra del dataframe:


,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
109004,PAYMENT,5,4,0.000000,210.000000,Late delivery,1,24,Women's Apparel,East Lansing,EE. UU.,XXXXXXXXX,Mary,5739,Palmer,XXXXXXXXX,Consumer,MI,1828 Iron Goose Mount,48823.0,5,Golf,42.772957,-84.435387,Africa,Nguigmi,Níger,5739,8/29/2016 20:52,41574,502,40.00,0.16,103778,50.000000,0.00,5,250.000000,210.000000,0.000000,West Africa,Diffa,PENDING_PAYMENT,NaN,502,24,NaN,http://images.acmesports.sports/Nike+Men%27s+D...,Nike Men's Dri-FIT Victory Golf Polo,50.000000,0,9/3/2016 20:52,Standard Class
1785,TRANSFER,6,4,37.930000,199.639999,Late delivery,1,37,Electronics,Riverside,EE. UU.,XXXXXXXXX,Mary,10731,Smith,XXXXXXXXX,Consumer,CA,6630 Sleepy Mews,92507.0,6,Outdoors,33.985275,-117.337707,Europe,London,Reino Unido,10731,9/5/2015 11:44,16955,821,8.32,0.04,42380,51.990002,0.19,4,207.960007,199.639999,37.930000,Northern Europe,Inglaterra,PROCESSING,NaN,821,37,NaN,http://images.acmesports.sports/Titleist+Pro+V...,Titleist Pro V1 High Numbers Personalized Gol,51.990002,0,9/11/2015 11:44,Standard Class
98492,DEBIT,4,4,54.410000,118.290001,Shipping on time,0,18,Men's Footwear,Chicago,EE. UU.,XXXXXXXXX,Kathy,8337,Jenkins,XXXXXXXXX,Home Office,IL,3114 Merry Gate Downs,60616.0,4,Apparel,41.895649,-87.704552,Africa,Owerri,Nigeria,8337,1/17/2017 8:35,51198,403,11.70,0.09,127947,129.990005,0.46,1,129.990005,118.290001,54.410000,West Africa,Imo,ON_HOLD,NaN,403,18,NaN,http://images.acmesports.sports/Nike+Men%27s+C...,Nike Men's CJ Elite 2 TD Football Cleat,129.990005,0,1/21/2017 8:35,Standard Class
55694,PAYMENT,2,2,79.680000,165.990005,Shipping on time,0,48,Water Sports,Caguas,Puerto Rico,XXXXXXXXX,Shirley,3119,Hinton,XXXXXXXXX,Corporate,PR,1927 Harvest Lane,725.0,7,Fan Shop,18.244633,-66.370514,LATAM,Lima,Perú,3119,3/26/2017 4:59,55846,1073,34.00,0.17,139644,199.990005,0.48,1,199.990005,165.990005,79.680000,South America,Lima (ciudad),PAYMENT_REVIEW,NaN,1073,48,NaN,http://images.acmesports.sports/Pelican+Sunstr...,Pelican Sunstream 100 Kayak,199.990005,0,3/28/2017 4:59,Second Class
73350,PAYMENT,2,4,-71.400002,167.990005,Advance shipping,0,48,Water Sports,Caguas,Puerto Rico,XXXXXXXXX,Mary,3584,Washington,XXXXXXXXX,Corporate,PR,5817 Green Hills Via,725.0,7,Fan Shop,18.296148,-66.370605,Pacific Asia,Rafsanjan,Irán,3584,12/6/2016 3:23,48306,1073,32.00,0.16,120821,199.990005,-0.43,1,199.990005,167.990005,-71.400002,South Asia,Kerman,PENDING_PAYMENT,NaN,1073,48,NaN,http://images.acmesports.sports/Pelican+Sunstr...,Pelican Sunstream 100 Kayak,199.990005,0,12/8/2016 3:23,Standard Class


In [6]:
df_data.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Cally,20755,Holloway,XXXXXXXXX,Consumer,PR,5365 Noble Nectar Island,725.0,2,Fitness,18.251453,-66.037056,Pacific Asia,Bekasi,Indonesia,20755,1/31/2018 22:56,77202,1360,13.110000,0.04,180517,327.75,0.29,1,327.75,314.640015,91.250000,Southeast Asia,Java Occidental,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Irene,19492,Luna,XXXXXXXXX,Consumer,PR,2679 Rustic Loop,725.0,2,Fitness,18.279451,-66.037064,Pacific Asia,Bikaner,India,19492,1/13/2018 12:27,75939,1360,16.389999,0.05,179254,327.75,-0.80,1,327.75,311.359985,-249.089996,South Asia,Rajastán,PENDING,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,EE. UU.,XXXXXXXXX,Gillian,19491,Maldonado,XXXXXXXXX,Consumer,CA,8510 Round Bear Gate,95125.0,2,Fitness,37.292233,-121.881279,Pacific Asia,Bikaner,India,19491,1/13/2018 12:06,75938,1360,18.030001,0.06,179253,327.75,-0.80,1,327.75,309.720001,-247.779999,South Asia,Rajastán,CLOSED,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,EE. UU.,XXXXXXXXX,Tana,19490,Tate,XXXXXXXXX,Home Office,CA,3200 Amber Bend,90027.0,2,Fitness,34.125946,-118.291016,Pacific Asia,Townsville,Australia,19490,1/13/2018 11:45,75937,1360,22.940001,0.07,179252,327.75,0.08,1,327.75,304.809998,22.860001,Oceania,Queensland,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Orli,19489,Hendricks,XXXXXXXXX,Corporate,PR,8671 Iron Anchor Corners,725.0,2,Fitness,18.253769,-66.037048,Pacific Asia,Townsville,Australia,19489,1/13/2018 11:24,75936,1360,29.500000,0.09,179251,327.75,0.45,1,327.75,298.250000,134.210007,Oceania,Queensland,PENDING_PAYMENT,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [7]:
df_data[df_data["Customer Id"] == 4598].head() #para corroborar que la latitud y longitud pertenecen al cliente, no a la tienda.

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
51273,DEBIT,5,4,19.670000,110.489998,Late delivery,1,32,Golf Apparel,Aurora,EE. UU.,XXXXXXXXX,Gloria,4598,Castro,XXXXXXXXX,Consumer,CO,7489 Stony End,80013.0,6,Outdoors,-33.937553,18.571438,Europe,Vallauris,Francia,4598,7/20/2017 5:31,63794,715,19.50,0.15,159455,129.990005,0.18,1,129.990005,110.489998,19.670000,Western Europe,Provenza-Alpes-Costa Azul,COMPLETE,NaN,715,32,NaN,http://images.acmesports.sports/TaylorMade+Wom...,TaylorMade Women's RBZ SL Rescue,129.990005,0,7/25/2017 5:31,Standard Class
94210,DEBIT,3,4,104.150002,371.980011,Advance shipping,0,45,Fishing,Aurora,EE. UU.,XXXXXXXXX,Gloria,4598,Castro,XXXXXXXXX,Consumer,CO,7489 Stony End,80013.0,7,Fan Shop,-33.937553,18.571438,Pacific Asia,Melbourne,Australia,4598,1/13/2016 19:16,25882,1004,28.00,0.07,64788,399.980011,0.28,1,399.980011,371.980011,104.150002,Oceania,Victoria,COMPLETE,NaN,1004,45,NaN,http://images.acmesports.sports/Field+%26+Stre...,Field & Stream Sportsman 16 Gun Fire Safe,399.980011,0,1/16/2016 19:16,Standard Class
112780,PAYMENT,3,4,77.370003,179.929993,Advance shipping,0,46,Indoor/Outdoor Games,Aurora,EE. UU.,XXXXXXXXX,Gloria,4598,Castro,XXXXXXXXX,Consumer,CO,7489 Stony End,80013.0,7,Fan Shop,-33.937553,18.571438,Pacific Asia,Rasht,Irán,4598,10/19/2016 18:52,45062,1014,19.99,0.10,112514,49.980000,0.43,4,199.919998,179.929993,77.370003,South Asia,Guilán,PAYMENT_REVIEW,NaN,1014,46,NaN,http://images.acmesports.sports/O%27Brien+Men%...,O'Brien Men's Neoprene Life Vest,49.980000,0,10/22/2016 18:52,Standard Class
128013,DEBIT,3,4,25.070000,75.980003,Advance shipping,0,29,Shop By Sport,Aurora,EE. UU.,XXXXXXXXX,Gloria,4598,Castro,XXXXXXXXX,Consumer,CO,7489 Stony End,80013.0,5,Golf,-33.937553,18.571438,Pacific Asia,Melbourne,Australia,4598,1/13/2016 19:16,25882,627,4.00,0.05,64787,39.990002,0.33,2,79.980003,75.980003,25.070000,Oceania,Victoria,COMPLETE,NaN,627,29,NaN,http://images.acmesports.sports/Under+Armour+G...,Under Armour Girls' Toddler Spine Surge Runni,39.990002,0,1/16/2016 19:16,Standard Class
128025,DEBIT,3,4,-4.500000,90.000000,Advance shipping,0,24,Women's Apparel,Aurora,EE. UU.,XXXXXXXXX,Gloria,4598,Castro,XXXXXXXXX,Consumer,CO,7489 Stony End,80013.0,5,Golf,-33.937553,18.571438,Pacific Asia,Melbourne,Australia,4598,1/13/2016 19:16,25882,502,10.00,0.10,64786,50.000000,-0.05,2,100.000000,90.000000,-4.500000,Oceania,Victoria,COMPLETE,NaN,502,24,NaN,http://images.acmesports.sports/Nike+Men%27s+D...,Nike Men's Dri-FIT Victory Golf Polo,50.000000,0,1/16/2016 19:16,Standard Class


In [8]:
df_data["latitude"].unique()

KeyError: 'latitude'

In [ ]:
df_data.groupby("Department Id")["Order Id"].value_counts().sort_values(ascending=False).head(10)

Department Id  Order Id
5              31242       5
7              58549       5
               5443        5
               25752       5
               26862       5
               55052       5
               56068       5
               52950       5
               33118       5
               25135       5
Name: count, dtype: int64

In [ ]:
df_data[df_data["Department Id"] == 5].sample(5) #para corroborar que la latitud y longitud pertenecen al cliente, no a la tienda.

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
29192,PAYMENT,1,0,98.000000,200.000000,Late delivery,1,24,Women's Apparel,Brooklyn,EE. UU.,XXXXXXXXX,Mary,12232,Moore,XXXXXXXXX,Corporate,NY,5735 Quaking Cider Highway,11216.0,5,Golf,40.679138,-73.944054,USCA,Fayetteville,Estados Unidos,12232,5/7/2016 21:44,33767,502,0.00,0.00,84270,50.000000,0.49,4,200.000000,200.000000,98.000000,South of USA,Carolina del Norte,PENDING_PAYMENT,28314.0,502,24,NaN,http://images.acmesports.sports/Nike+Men%27s+D...,Nike Men's Dri-FIT Victory Golf Polo,50.000000,0,5/8/2016 9:44,Same Day
31333,TRANSFER,2,4,23.590000,212.500000,Advance shipping,0,24,Women's Apparel,Caguas,Puerto Rico,XXXXXXXXX,Madison,3188,Ray,XXXXXXXXX,Home Office,PR,9276 Jagged Towers,725.0,5,Golf,18.297707,-66.370560,LATAM,Juárez,México,3188,6/12/2017 21:20,61236,502,37.50,0.15,153168,50.000000,0.11,5,250.000000,212.500000,23.590000,Central America,Chihuahua,PENDING,NaN,502,24,NaN,http://images.acmesports.sports/Nike+Men%27s+D...,Nike Men's Dri-FIT Victory Golf Polo,50.000000,0,6/14/2017 21:20,Standard Class
117517,TRANSFER,0,0,6.800000,75.580002,Shipping on time,0,29,Shop By Sport,Wheaton,EE. UU.,XXXXXXXXX,Kenneth,4432,Gillespie,XXXXXXXXX,Consumer,IL,4294 High Passage,60187.0,5,Golf,41.865440,-88.109901,Pacific Asia,Visakhapatnam,India,4432,12/3/2015 7:35,23040,627,4.40,0.06,57693,39.990002,0.09,2,79.980003,75.580002,6.800000,South Asia,Andhra Pradesh,PENDING,NaN,627,29,NaN,http://images.acmesports.sports/Under+Armour+G...,Under Armour Girls' Toddler Spine Surge Runni,39.990002,0,12/3/2015 19:35,Same Day
56950,DEBIT,4,4,17.719999,141.750000,Shipping on time,0,24,Women's Apparel,Caguas,Puerto Rico,XXXXXXXXX,Mary,4761,Shannon,XXXXXXXXX,Consumer,PR,8289 Emerald Pike,725.0,5,Golf,18.244875,-66.370598,USCA,Kent,Estados Unidos,4761,6/16/2016 16:46,36493,502,8.25,0.06,91117,50.000000,0.13,3,150.000000,141.750000,17.719999,West of USA,Washington,COMPLETE,98031.0,502,24,NaN,http://images.acmesports.sports/Nike+Men%27s+D...,Nike Men's Dri-FIT Victory Golf Polo,50.000000,0,6/20/2016 16:46,Standard Class
28984,PAYMENT,1,0,21.600000,45.000000,Late delivery,1,24,Women's Apparel,Bronx,EE. UU.,XXXXXXXXX,Mary,6156,Hernandez,XXXXXXXXX,Corporate,NY,5918 Golden Beacon Freeway,10461.0,5,Golf,40.846256,-73.847252,LATAM,Buenos Aires,Argentina,6156,5/18/2017 17:00,59511,502,5.00,0.10,148951,50.000000,0.48,1,50.000000,45.000000,21.600000,South America,Buenos Aires,PENDING_PAYMENT,NaN,502,24,NaN,http://images.acmesports.sports/Nike+Men%27s+D...,Nike Men's Dri-FIT Victory Golf Polo,50.000000,0,5/19/2017 5:00,Same Day


In [ ]:
df_data.groupby("Customer Id")["Order Id"].value_counts().sort_values(ascending=False).head(10)

Customer Id  Order Id
2            57963       5
4995         58897       5
4996         36451       5
             6717        5
4997         63047       5
4987         45977       5
             30412       5
4989         11862       5
12425        48314       5
12426        32304       5
Name: count, dtype: int64

In [ ]:
df_data["shipping date (DateOrders)"].value_counts()

shipping date (DateOrders)
5/9/2015 18:02      10
9/27/2015 20:28     10
9/24/2015 23:48     10
8/1/2015 18:37      10
11/15/2016 13:45    10
                    ..
1/27/2016 18:59      1
1/28/2016 19:25      1
1/26/2016 6:49       1
1/23/2016 7:10       1
1/21/2016 5:56       1
Name: count, Length: 63701, dtype: int64

In [ ]:
df_data["Shipping Mode"].value_counts()

Shipping Mode
Standard Class    107752
Second Class       35216
First Class        27814
Same Day            9737
Name: count, dtype: int64

In [ ]:
df_data["Customer Country"].value_counts()

Customer Country
EE. UU.        111146
Puerto Rico     69373
Name: count, dtype: int64

In [ ]:
df_data["Customer State"].unique()

<StringArray>
[   'PR',    'CA',    'NY',    'FL',    'MA',    'IL',    'MT',    'PA',
    'MI',    'TX',    'DE',    'GA',    'MD',    'OH',    'HI',    'NJ',
    'WI',    'AZ',    'CO',    'MN',    'NC',    'NM',    'OR',    'SC',
    'VA',    'UT',    'WA',    'KY',    'WV',    'RI',    'CT',    'LA',
    'TN',    'DC',    'ND',    'MO',    'IN',    'ID',    'NV',    'KS',
    'AR',    'OK',    'AL',    'IA', '95758', '91732']
Length: 46, dtype: str

In [ ]:
df_data["Order Status"].unique()

<StringArray>
[       'COMPLETE',         'PENDING',          'CLOSED', 'PENDING_PAYMENT',
        'CANCELED',      'PROCESSING', 'SUSPECTED_FRAUD',         'ON_HOLD',
  'PAYMENT_REVIEW']
Length: 9, dtype: str

In [ ]:
df_data["Order Region"].unique()

<StringArray>
[ 'Southeast Asia',      'South Asia',         'Oceania',    'Eastern Asia',
       'West Asia',    'West of USA ',      'US Center ',     'West Africa',
  'Central Africa',    'North Africa',  'Western Europe', 'Northern Europe',
 'Central America',       'Caribbean',   'South America',     'East Africa',
 'Southern Europe',     'East of USA',          'Canada', 'Southern Africa',
    'Central Asia',  'Eastern Europe',  'South of  USA ']
Length: 23, dtype: str

In [ ]:
df_data["Market"].unique()

<StringArray>
['Pacific Asia', 'USCA', 'Africa', 'Europe', 'LATAM']
Length: 5, dtype: str

In [ ]:
df_data["Category Name"].unique()

<StringArray>
[      'Sporting Goods',               'Cleats',        'Shop By Sport',
      'Women's Apparel',          'Electronics',         'Boxing & MMA',
     'Cardio Equipment',             'Trade-In',     'Kids' Golf Clubs',
   'Hunting & Shooting',  'Baseball & Softball',       'Men's Footwear',
     'Camping & Hiking', 'Consumer Electronics',             'Cameras ',
            'Computers',           'Basketball',               'Soccer',
       'Girls' Apparel',          'Accessories',     'Women's Clothing',
               'Crafts',       'Men's Clothing',     'Tennis & Racquet',
  'Fitness Accessories',      'As Seen on  TV!',           'Golf Balls',
    'Strength Training',  'Children's Clothing',             'Lacrosse',
                'Baby ',              'Fishing',               'Books ',
                 'DVDs',                 'CDs ',               'Garden',
               'Hockey',         'Pet Supplies',    'Health and Beauty',
                'Music',          'Vi

In [ ]:
df_data["Customer Segment"].unique()

<StringArray>
['Consumer', 'Home Office', 'Corporate']
Length: 3, dtype: str

In [ ]:
df_data.groupby("Customer State")["Order Zipcode"].unique()

Customer State
91732                                                [nan]
95758                                                [nan]
AL                [nan, 60462.0, 98105.0, 7960.0, 22980.0]
AR               [nan, 43130.0, 90032.0, 39212.0, 43302.0]
AZ       [nan, 10035.0, 32216.0, 43229.0, 46368.0, 8001...
CA       [nan, 95123.0, 19134.0, 31907.0, 77041.0, 2230...
CO       [nan, 26003.0, 38109.0, 92024.0, 19134.0, 1002...
CT       [nan, 89115.0, 19140.0, 95051.0, 90049.0, 7501...
DC       [nan, 73120.0, 31907.0, 48227.0, 47201.0, 9203...
DE       [nan, 43055.0, 60505.0, 90049.0, 90032.0, 7050...
FL       [nan, 38109.0, 72401.0, 60610.0, 6708.0, 43017...
GA       [nan, 10035.0, 97206.0, 98661.0, 87124.0, 7570...
HI       [60016.0, nan, 75034.0, 13440.0, 89031.0, 1000...
IA               [nan, 10011.0, 85023.0, 48911.0, 78041.0]
ID               [nan, 38301.0, 48234.0, 20735.0, 19711.0]
IL       [nan, 13021.0, 79762.0, 32216.0, 49201.0, 9811...
IN       [nan, 37042.0, 60623.0, 94110.0,

In [ ]:
df_token = pd.read_csv("../files/raw/tokenized_access_logs.csv")

In [ ]:
e.info_df(df_token)


INFORME TÉCNICO DEL DATAFRAME

🔹 Dimensiones:
- Filas: 469977
- Columnas: 8

🔹 Info general:
<class 'pandas.DataFrame'>
RangeIndex: 469977 entries, 0 to 469976
Data columns (total 8 columns):
 #   Column      Non-Null Count   Dtype
---  ------      --------------   -----
 0   Product     469977 non-null  str  
 1   Category    469977 non-null  str  
 2   Date        469977 non-null  str  
 3   Month       469977 non-null  str  
 4   Hour        469977 non-null  int64
 5   Department  469977 non-null  str  
 6   ip          469977 non-null  str  
 7   url         469977 non-null  str  
dtypes: int64(1), str(7)
memory usage: 28.7 MB

🔹 Nulos:
Product       0.0
Category      0.0
Date          0.0
Month         0.0
Hour          0.0
Department    0.0
ip            0.0
url           0.0
dtype: float64

🔹 Duplicados:
Total duplicados: 3249

🔹 Descriptivo variables cualitativas:


,count,unique,top,freq
Product,469977,76,Perfect Fitness Perfect Rip Deck,27878
Category,469977,33,cleats,27878
Date,469977,160815,9/14/2017 20:34,156
Month,469977,5,Sep,137238
Department,469977,6,outdoors,79926
ip,469977,3340,157.21.93.193,566
url,469977,152,/department/apparel/category/cleats/product/Pe...,20258



🔹 Descriptivo variables cuantitativas:


,count,mean,std,min,25%,50%,75%,max
Hour,469977.0,14.59,5.57,0.0,10.0,15.0,20.0,23.0



🔹 Muestra del dataframe:


,Product,Category,Date,Month,Hour,Department,ip,url
351058,Perfect Fitness Perfect Rip Deck,cleats,1/12/2018 9:02,Jan,9,apparel,173.219.18.85,/department/apparel/category/cleats/product/Pe...
4948,adidas Men's Germany Black Crest Away Tee,girls' apparel,10/10/2017 15:38,Oct,15,golf,72.217.87.201,/department/golf/category/girls'%20apparel/pro...
148764,Nike Men's Fingertrap Max Training Shoe,soccer,11/4/2017 10:20,Nov,10,fitness,185.141.25.207,/department/fitness/category/soccer/product/Ni...
178603,Diamondback Boys' Insight 24 Performance Hybr,basketball,12/20/2017 22:47,Dec,22,fitness,88.197.205.228,/department/fitness/category/basketball/produc...
22630,Nike Men's Fingertrap Max Training Shoe,soccer,1/31/2018 18:15,Jan,18,fitness,26.34.173.88,/department/fitness/category/soccer/product/Ni...


In [ ]:
df_token["Month"].unique()

<StringArray>
['Sep', 'Oct', 'Dec', 'Jan', 'Nov']
Length: 5, dtype: str

# Conclusiones finales del EDA · DataCo
 
- **Calidad de Datos y Gestión de Nulos:** La integridad general del dataset es excelente en la mayoría de sus variables transaccionales y financieras. La excepción principal se concentra en la columna `Order Zipcode`, la cual presenta valores nulos que requieren una estrategia de imputación o limpieza en la fase ETL, dado que pueden afectar el análisis geoespacial del destino de las órdenes.
 
- **Duplicados y Estructura Lógica:** Las repeticiones detectadas en los registros no constituyen errores de duplicidad aleatoria; se justifican plenamente por la estructura lógica y granular del dataset, donde un mismo número de orden (`Order Id`) puede contener múltiples ítems de productos individuales o registrar distintas actualizaciones en el historial de envío.
 
- **Políticas de Privacidad y Reducción de Dimensiones (Eliminación):** Con el fin de optimizar el rendimiento del almacenamiento en el ecosistema de Big Data y garantizar el cumplimiento estricto de las normativas de protección de datos (debido a la naturaleza sintética/expuesta de cierta información), se determina la eliminación definitiva de las siguientes columnas no relevantes o sensibles:
  - `Customer Email` (Privacidad)
  - `Customer Password` (Seguridad)
  - `Product Description` (Columna redundante o vacía en el catálogo de productos)
  - `Order State` (Duplicidad con las métricas regionales o de envío)
 
- **Estandarización y Corrección de Anomalías:** Se detectaron inconsistencias críticas en la columna de ubicación del cliente (`Customer State`). Mientras que la mayoría de los registros adoptan el formato estándar de código de estado de dos letras (ej. NY, TX), se identificaron códigos postales numéricos (como *95758* y *91732*) en dicho campo. Como acción obligatoria de limpieza, se plantea mapear y reemplazar estos números por el estado correcto al que pertenecen (`CA`).
 
- **Optimización de Estructura y Formato de Columnas:** Dado que las columnas ya se encuentran en el idioma inglés, la transformación se centrará en estandarizar su formato de texto. Se aplicará una conversión estricta a minúsculas y se reemplazarán los espacios por guiones bajos (formato `snake_case`) para homogeneizar los nombres de las variables (ej. de `Order Id` a `order_id`). Asimismo, se transformarán las variables temporales desde texto plano a formato `datetime` nativo.
 
- **Integración y Complementariedad:** Las tablas analizadas demuestran ser altamente complementarias para construir un modelo en estrella o copo de nieve, permitiendo cruzar de manera directa las dimensiones de Clientes, Productos y Geografía con la tabla de hechos de Órdenes y Envíos.
 
---
 
Esta fase concluye con éxito, dejando una hoja de ruta clara, justificada y accionable para la siguiente etapa del pipeline: **Limpieza, Transformación y Carga (ETL)**.